File: scrapers/limetorrents.js
const express = require("express");
const router = express.Router();
const axios = require("axios");
const cheerio = require("cheerio");
//  Base URL of skytorrents
const BASE_URL = require("./constants").LIMETORRENTS_BASE_URL;
const isMagnet = require("./utils/misc_utils.js").isMagnet;

router.get("/limetorrents", async function (req, res) {
    //  Get the Item to be searched from Query
    var search = req.query.search.trim();
    response = await axios.get(BASE_URL + search).catch((err) => {
        console.log(err);
    });
    var $ = cheerio.load(response.data);
    var jsonResponse = [];
    var selector = $("table.table2 tbody tr");
    if (selector.length > 1) {
        selector.each((index, element) => {
            //  File Name
            name = $(element).find("td:nth-child(1)").text();
            //  Seeders
            seeders = $(element).find("td:nth-child(4)").text();
            //  Leechers
            leechers = $(element).find("td:nth-child(5)").text();
            //  Upload Date
            upload_date = $(element)
                .find("td:nth-child(2)")
                .text()
                .split(" - ")[0];
            //  File Size
            file_size = $(element).find("td:nth-child(3)").text();

            //  magnet
            magnet = "";
            url =
                "https://www.limetorrents.info" +
                $(element)
                    .find("td.tdleft div.tt-name a:nth-child(2)")
                    .attr("href");
            jsonResponse.push({
                name: name,
                torrent_url: url,
                seeders: seeders,
                leechers: leechers,
                upload_date: upload_date,
                size: file_size,
                uploader: "--",
                magnet: magnet,
                website: "Limetorrents",
            });
        });
        jsonResponse.shift();
        res.status(200).json({ data: jsonResponse }).end();
    } else {
        res.status(204).end();
    }
});

router.get("/limetorrents_mg", async function (req, res) {
    url = req.query.url;
    response = await axios.get(url).catch((err) => {
        console.log(err);
    });
    $ = cheerio.load(response.data);
    magnet = $(
        "#content > div:nth-child(6) > div:nth-child(1) > div > div:nth-child(13) > div > p > a"
    ).attr("href");
    if (isMagnet(magnet)) {
        res.status(200).json({ magnet: magnet }).end();
    } else {
        res.status(204).end();
    }
});

module.exports = router;

File: routes/Limetorrents.go
package routes

import (
	"fmt"
	"strings"

	"github.com/gin-gonic/gin"
	"github.com/gocolly/colly/v2"
)

func Limetorrents(ginCon *gin.Context) {
	search := strings.ReplaceAll(strings.TrimSpace(ginCon.Query("search")), " ", "%20")
	url := fmt.Sprintf("https://www.limetorrents.info/search/all/%s/", search)
	c := colly.NewCollector()
	infos := make([]TorrentInfo, 0)
	c.OnHTML("body", func(e *colly.HTMLElement) {
		ti := TorrentInfo{}
		e.ForEach("table.table2 tbody tr", func(i int, e *colly.HTMLElement) {
			if i == 0 {
				return
			}
			ti.Name = e.ChildText("td:nth-child(1)")
			ti.Seeders = e.ChildText("td:nth-child(4)")
			ti.Leechers = e.ChildText("td:nth-child(5)")
			ti.Date = strings.Split(e.ChildText("td:nth-child(2)"), " - ")[0]
			ti.Size = e.ChildText("td:nth-child(3)")
			ti.Magnet = ""
			ti.Url = "https://www.limetorrents.info" + e.ChildAttr("td.tdleft div.tt-name a:nth-child(2)", "href")
			ti.Website = "Limetorrents"
			ti.Uploader = "N/A"
			ti.TorrentFileUrl = ""
			infos = append(infos, ti)
		})
	})
	c.OnError(func(r *colly.Response, err error) {
		fmt.Println("Request URL:", r.Request.URL, "failed with response:", r, "\nError:", err)
	})
	c.OnScraped(func(r *colly.Response) {
		if len(infos) > 0 {
			ginCon.JSON(200, TorrentRepo{infos})
		} else {
			ginCon.AbortWithStatus(204)
		}
	})
	c.Visit(url)
}

func Limetorrents_getMagnet(ginCon *gin.Context) {
	url := ginCon.Query("url")
	c := colly.NewCollector()
	var magnet, torrentfile string
	c.OnHTML("body", func(e *colly.HTMLElement) {
		magnet = e.ChildAttr("body > div > table:nth-child(6) > tbody > tr:nth-child(5) > td > span > a", "href")
		torrentfile = e.ChildAttr("body > div > table:nth-child(6) > tbody > tr:nth-child(3) > td > span > a", "href")
	})

	c.OnError(func(r *colly.Response, err error) {
		fmt.Println("Request URL:", r.Request.URL, "failed with response:", r, "\nError:", err)
	})

	c.OnScraped(func(r *colly.Response) {
		if strings.HasPrefix(magnet, "magnet") {
			ginCon.JSON(200, gin.H{"magnet": magnet, "torrentFile": torrentfile})
		} else {
			ginCon.AbortWithStatus(204)
		}
	})
	c.Visit(url)
}
